ds3_java_cli/src/main/java/com/spectralogic/ds3cli/command/GetBulk.java

```diff
@@ -98,81 +114,76 @@ public CliCommand init(final Arguments args) throws Exception {
     public DefaultResult call() throws Exception {
-        if (Guard.isStringNullOrEmpty(this.directory) || directory.equals(".")) {
-            this.outputPath = FileSystems.getDefault().getPath(".");
-        } else {
-            final Path dirPath = FileSystems.getDefault().getPath(directory);
-            this.outputPath = FileSystems.getDefault().getPath(".").resolve(dirPath);
-        }
-
-        final Ds3ClientHelpers.ObjectChannelBuilder getter;
-        if (this.checksum) {
-           throw new CommandException("Checksumming is currently not implemented.");
-
-//          TOD):  Logging.log("Performing get_bulk with checksum verification");
-//            getter = new VerifyingFileObjectGetter(this.outputPath);
-        } else if (this.discard) {
-            LOG.warn("Using /dev/null getter -- all incoming data will be discarded");
-            getter = new MemoryObjectChannelBuilder(DEFAULT_BUFFER_SIZE, DEFAULT_FILE_SIZE);
-        } else {
-            getter = new FileObjectGetter(this.outputPath);
-        }
+        final Ds3ClientHelpers.ObjectChannelBuilder getter
+                = this.discard ?  new MemoryObjectChannelBuilder(DEFAULT_BUFFER_SIZE, DEFAULT_FILE_SIZE)
+                : new FileObjectGetter(this.outputPath);
 
         if (this.sync) {
-            if (Guard.isStringNullOrEmpty(this.prefix)) {
+            if (Guard.isNullOrEmpty(this.prefixes)) {
                 LOG.info("Syncing all objects from {}", this.bucketName);
             } else {
-                LOG.info("Syncing only those objects that start with {}", this.prefix);
+                LOG.info("Syncing only those objects that start with {}", Joiner.on(PREFIX_SEPARATOR).join(this.prefixes));
             }
             return new DefaultResult(this.restoreSome(getter));
         }
 
-        if (Guard.isStringNullOrEmpty(this.prefix)) {
+        if (!Guard.isNotNullAndNotEmpty(prefixes)) {
             LOG.info("Getting all objects from {}", this.bucketName);
             return new DefaultResult(this.restoreAll(getter));
         }
 
-        LOG.info("Getting only those objects that start with {}", this.prefix);
+        LOG.info("Getting only those objects that start with {}", Joiner.on(PREFIX_SEPARATOR).join(this.prefixes));
         return new DefaultResult(this.restoreSome(getter));
     }
 
     private String restoreSome(final Ds3ClientHelpers.ObjectChannelBuilder getter) throws IOException, XmlProcessingException {
         final Ds3ClientHelpers helper = getClientHelpers();
-        final Iterable<Contents> contents = helper.listObjects(this.bucketName, this.prefix);
 
-        final Iterable<Contents> filteredContents;
+        Iterable<Contents> prefixMatches;
+        if (Guard.isNullOrEmpty(prefixes)) {
+            prefixMatches = new LazyIterable<>(
+                    new GetBucketLoaderFactory(getClient(), this.bucketName, null, null, 100, 5));
+        } else {
+            prefixMatches = new ArrayList<Contents>();
```

Comment:
I would re-write this section like this:

```java
Iterable<Contents> prefixedIterable = Collections.emptyList();
for (final String prefix : prefixes) {
    final Iterable<Contents> prefixMatch = new LazyIterable<>(
                         new GetBucketLoaderFactory(getClient(), this.bucketName, prefix, null, 100, 5));
    prefixedIterable = Iterables.concat(prefixedIterable, prefixMatch);
}
```

And then when it's moved into it's own method, you can return `prefixedIterable`